text/0000-autoref-copy.md

```diff
@@ -0,0 +1,200 @@
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+When working with `Copy` data, the distinction between borrowed and owned data
+is often unimportant. However, generic code often results in code which will
+only accept a particular variant of a `Copy` type (`T`, `&T`, `&&T`, ...).
+This is a frustration in many places:
+
+```rust
+// Comparisons:
+let v = vec![0, 1, 2, 3];
+v.iter().filter(|x| x > 1); // ERROR: expected &&_, found integral variable
+// These work:
+0 > 1;
+&0 > &1;
+// But these don't:
+&0 > 1;
+0 > &1;
+
+// Trait instantiations:
+let mut map: HashMap::new();
+map.insert(0, "Hello");
+map.insert(1, "world!");
+map[0]; // ERROR: expected &{integer}, found integral variable
+// or
+map.get(1); // ERROR: expected `&{integer}`, found integral variable
+
+// Numeric operators:
+// These work:
+&0 + &1;
+0 + &1;
+&0 + 1;
+// But these don't:
+&&0 + 1;
+&&0 + &1;
+&&0 + &&1;
+```
+
+These interactions confuse both new and experienced users without providing
+any significant value. It's clear what's intended by `map.get(1)` or
+`vec.iter().filter(|x| x > 1)`. When users encounter these errors in practice,
+the only reasonable thing they can do is add `&` and `*` as necessary to make
+their code compile.
+
+This RFC seeks to address one particular variant of this problem: passing
+owned `Copy` data where a reference was expected.
+
+Example:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    println!("{:?}", map.get(1)); // ERROR: expected `&{integer}`
+
+    // Instead, users have to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+This is an unnecessary frustration. This RFC proposes to prevent this issue
+by auto-referencing `Copy` types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When calling a function which expects a reference to a `Copy` type, you can
+pass an owned `Copy` type and a reference will be created automatically:
+
+```rust
+fn print_num(x: &usize) { println!("{:?}", x); }
+
+fn main() {
+  let x = 5;
+  print_num(x);
+}
+```
+
+This is particularly convenient when working with some generic types whose
+methods take references as arguments:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    // We can write this:
+    println!("{:?}", map.get(1));
+
+    // Instead of having to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`Copy` types will autoreference: at
+[coercion sites](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md#coercions),
+when a reference is expected, but an owned
+`Copy` type is found, an `&` will be automatically inserted.
+
+If the `Copy` value is a temporary, its lifetime will be promoted to meet the
+required lifetime of the reference (where possible).
```

Comment:
I find this sentence at odds with the behavior specific in RFC 66 etc. Specifically, the sentence as written suggests to me that we will infer the appropriate lifetime for the temporary, which would be quite at odds with the rest of the language. I think I would say it more like this:

If the value to be borrowed is an rvalue (e.g., a constant like `1`), then a temporary will be created for it, just as if one explicitly wrote the `&` (e.g., `&1`). This temporary will have the same lifetime that would result from an explicit `&`, which generally means that it lives until the end of the innermost enclosing statement (until the next `;`) -- but in some cases it may live longer. See RFC 66 and @nikomatsakis's amendment for more details.

Note that a consequence of this is that in some cases there will be implicit borrows that might be surprising. An example:

```rust
fn foo(x: &i32) -> &i32 { x }

let mut x = 22
let p = foo(x); // implicitly: borrows `x`
x += 1; // ERROR -- `x` is borrowed
print(p);
```